reverse-proxy/src/tls_conf.rs

```diff
@@ -0,0 +1,144 @@
+
+        // provide the private key
+        {
+            let key = PKey::private_key_from_pem(server_key)
+                .inspect_err(|e| {
+                    log::error!("Failed to parse server private key: {}", e);
+                })
+                .unwrap();
+            ssl.set_private_key(&key)
+                .inspect_err(|e| {
+                    log::error!("Failed to set server private key: {}", e);
+                })
+                .unwrap();
+        }
+
+        // provide the certificate chain file
+        {
+            let cert = boring::x509::X509::from_pem(server_pem)
+                .inspect_err(|e| {
+                    log::error!("Failed to parse server certificate: {}", e);
+                })
+                .unwrap();
+
+            ssl.set_certificate(&cert)
+                .inspect_err(|e| {
+                    log::error!("Failed to set server certificate: {}", e);
+                })
+                .unwrap();
+        }
+
+        // add the CA certificate to the SSL context
+        let ca_cert = boring::x509::X509::from_pem(CA_PEM)
+            .inspect_err(|e| {
+                log::error!("Failed to parse CA certificate: {}", e);
+            })
+            .unwrap();
+        {
+            ssl.add_chain_cert(&ca_cert)
+                .inspect_err(|e| {
+                    log::error!("Failed to add CA certificate: {}", e);
+                })
+                .unwrap();
+        }
+
+        ssl.set_custom_verify_callback(
+            SslVerifyMode::PEER,
+            Self::verify_callback(ca_cert.public_key().unwrap()),
+        );
+    }
+}
+
+impl TlsConfig {
+    fn verify_callback(
+        ca_cert_pub_key: PKey<Public>,
+    ) -> Box<dyn Fn(&mut SslRef) -> Result<(), SslVerifyError> + 'static + Sync + Send> {
+        Box::new(move |ssl| -> Result<(), SslVerifyError> {
+            Self::verify_client_file(&ca_cert_pub_key, ssl)
+        })
+    }
+
+    fn verify_client_file(
+        server_ca_public_key: &PKey<Public>,
+        ssl: &mut TlsRef,
+    ) -> Result<(), SslVerifyError> {
+        if ssl.verify_mode() != SslVerifyMode::PEER {
+            log::error!("SSL verify mode is not set to PEER, cannot verify client certificate");
+            return Err(SslVerifyError::Invalid(SslAlert::INTERNAL_ERROR));
+        }
+
+        let client_cert = match ssl.peer_certificate() {
+            Some(val) => val,
+            None => {
+                log::error!("Failed to get client certificate");
+                return Err(SslVerifyError::Invalid(SslAlert::NO_CERTIFICATE));
+            }
+        };
+
+        // Making sure the client CN is "forward_proxy"
+        debug!("Debug Client certificate: {:?}", client_cert.subject_name());
+
+        // On the client side, the chain includes the leaf certificate, but on the server side it does
+        // not. Fun!
+        let client_ca_cert = match ssl.peer_cert_chain() {
+            Some(chain) => chain.get(0).unwrap(),
```

Comment:
~~Before returning, can you console out the chain? I bet it has length 2 and the ca of idx 0 is the root ca (i.e., ca.pem). If yes, than instead of the equality check here:~~
```
        // We expect the client CA public key to be the same as the one we have
        if !client_ca_pub_key.public_eq(server_ca_public_key) {
            log::error!("Client CA public key does not match server CA public key");
            return Err(SslVerifyError::Invalid(SslAlert::BAD_CERTIFICATE));
        }
```
~~Is we use the ca.pem public key loaded into the RP to check the signature of the FP ,`forward-proxy.pem` which i suspect comes through as `chain.get(1)`?~~

Reply:
Never mind. Great work! Understanding more now!